### PR TITLE
expression: derive collation and charset information and support the Collation() builtin-func

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve? 
Derive collation and charset information for builtin functions when creating them and support the `Collation()` builtin-function.

### What is changed and how it works?
1. derive collation and charset information when creating builtin functions according to this [doc](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-coercibility.html);
2. support the `Collation()` builtin-function;
3. add some integration tests.

### Check List 

Tests 

 - Unit test

### CODE REVIEW

File: expression/collation.go [hunk 1]
```diff
@@ -14,6 +14,9 @@
 package expression
 
 import (
+	"github.com/pingcap/tidb/sessionctx"
```
Comment (wjhuang2016) [inline expression/collation.go]: fmt

Comment (qw4990) [reply expression/collation.go]: fixed, PTAL

**Comment by qw4990** [comment]
PTAL @winoros @bb7133

**Pull request merged**